### PR TITLE
Use hashed assets for all env

### LINKED ISSUE #644: Dev server proxy bypassed by static asset servers
When using a local app server like [puma-dev](https://github.com/puma/puma-dev) or [pow](https://github.com/basecamp/pow) that serves static assets, starting `bin/webpack-dev-server` will have no effect if a compiled .js file already exists. The preexisting files in `public/packs` will be served directly and the requests never hit Rails.

Steps to reproduce:
1. Load a page with webpacker's `compile: true` option enabled (the default)
2. Start `bin/webpack-dev-server`
3. Change your application's js in some way and reload the page

/cc @gauravtiwari @dhh

### PR DESCRIPTION
**Pull request by gauravtiwari**
Fixes #644 #550

This PR enables hashed assets for all environments so we can expire them when changed

### CODE REVIEW

**Comment by javan** [comment]
Can remove `output: { filename: '[name]-[chunkhash].js' }` in `production.js` too, right?

File: lib/install/config/webpack/shared.js [hunk 1]
```diff
@@ -26,7 +26,7 @@ module.exports = {
   ),
 
   output: {
-    filename: '[name].js',
+    filename: '[name]-[hash].js',
```
Comment (javan) [inline lib/install/config/webpack/shared.js]: Should this be `[name]-[chunkhash].js`? That's what `production.js` uses.

Comment (gauravtiwari) [inline lib/install/config/webpack/shared.js]: Yeah but not available with dev server and watcher actually, just production build

Comment (justin808) [inline lib/install/config/webpack/shared.js]: @gauravtiwari do you have a doc link for that?

Comment (gauravtiwari) [inline lib/install/config/webpack/shared.js]: This is not properly explained in doc but basically chunkhash isn't available with HMR
screen shot 2017-08-16 at 09 56 25

**Pull request merged**